src/OrleansRuntime/GrainDirectory/GrainDirectoryPartition.cs

```diff
@@ -332,19 +332,22 @@ internal void RemoveGrain(GrainId grain)
         internal AddressesAndTag LookUpGrain(GrainId grain)
         {
             var result = new AddressesAndTag();
+            IGrainInfo grainInfo;
             lock (lockable)
             {
-                if (partitionData.ContainsKey(grain))
+                if (!partitionData.TryGetValue(grain, out grainInfo))
                 {
-                    result.Addresses = new List<ActivationAddress>();
-                    result.VersionTag = partitionData[grain].VersionTag;
-
-                    foreach (var route in partitionData[grain].Instances.Where(route => IsValidSilo(route.Value.SiloAddress)))
-                    {
-                        result.Addresses.Add(ActivationAddress.GetAddress(route.Value.SiloAddress, grain, route.Key, route.Value.RegistrationStatus));
-                    }
+                    return result;
                 }
             }
+
+            result.Addresses = new List<ActivationAddress>();
+            result.VersionTag = grainInfo.VersionTag;
+            foreach (var route in grainInfo.Instances.Where(route => IsValidSilo(route.Value.SiloAddress)))
```

Comment:
hmmm, I'm not sure this is safe, since now the dictionary returned by grainInfo.Instances can be mutated concurrently (see that updates to it are done within a lock).
You'd have to either clone into a list on read (which was what you were initially trying to avoid in this PR, but now it would be for an actually valid reason), or you'd have to change every update of the dictionary into a clone, update and swap operation (which could be trickier to understand the side effects of).
As I mentioned, I'm fine if we don't change this in this PR, and revert to the filtering being done within the lock for now. Still your initial commit makes sense regardless of this additional contention optimization.